packages/story-editor/src/app/canvas/useCanvasCopyPaste.js

```diff
@@ -103,20 +104,48 @@ function useCanvasGlobalKeys() {
         currentPage
       );
 
-      for (const { type, product } of elements) {
-        if (
-          type === ELEMENT_TYPES.PRODUCT &&
-          product?.productId &&
-          currentPageProductIds.includes(product.productId)
-        ) {
-          showSnackbar({
-            message: sprintf(
-              /* translators: %s: product title. */
-              __('Product "%s" already exists on the page.', 'web-stories'),
-              product.productTitle
+      const newProductsFromElements = elements
+        .filter(
+          ({ type, product }) =>
+            type === ELEMENT_TYPES.PRODUCT && product?.productId
+        )
+        .map(({ product }) => product);
+
+      if (
+        currentPageProductIds.length >= MAX_PRODUCTS_PER_PAGE ||
+        newProductsFromElements.length + currentPageProductIds.length >
+          MAX_PRODUCTS_PER_PAGE
+      ) {
+        showSnackbar({
+          message: sprintf(
+            /* translators: %d: max number of products. */
+            _n(
+              'Only %d item can be added per page.',
+              'Only %d items can be added per page.',
+              MAX_PRODUCTS_PER_PAGE,
+              'web-stories'
             ),
-          });
-        }
+            MAX_PRODUCTS_PER_PAGE
+          ),
+        });
+      } else {
+        newProductsFromElements.forEach(
+          ({ productId, productTitle, productImages }) => {
+            if (currentPageProductIds.includes(productId)) {
+              showSnackbar({
+                message: sprintf(
+                  /* translators: %s: product title. */
+                  __('Product "%s" already exists on the page.', 'web-stories'),
+                  productTitle
+                ),
+                thumbnail: productImages && {
+                  src: productImages[0].url,
+                  alt: productImages[0].alt,
+                },
```

Comment:
I know we currently disallow products without images, but if this ever changes and `productImages` is an empty array here, this will silently fail.

This adds some hardening:

```suggestion
                thumbnail: productImages?.[0]?.url && {
                  src: productImages[0].url,
                  alt: productImages[0].alt,
                },
```

Reply:
Done.